drivers: mipi_dbi: add support for parallel 8080/6800 modes using GPIO

stgloorious opened the pull request:
Introduce GPIO-based driver for MIPI DBI class that allows MIPI DBI type A and B displays to be used on general platforms. Until now, only Type C (SPI-based) MIPI DBI displays and some SoC-specific peripherals are supported for the mipi-dpi driver class.

Since each data pin GPIO can be selected individually in the device tree, the data bus pins are set in a loop for each transmitted byte. Obviously this has a significant negative impact on performance, however, it enables maximum flexbility and compatibility with all kinds of systems.

When using 8-bit mode and all the data GPIO pins are on the same port, a data port look-up table is generated automatically to set the whole port at once as a performance optimization. This creates a RAM overhead of about 1 kiB, and is only implemented for 8-bit mode as the table grows exponentially with bus width. To me, it seems like a reasonable tradeoff between RAM overhead and performance as MCUs beefy enough to run a such a TFT display usually have enough RAM to spare.

I tested the 8-bit 8080 mode with a ILI9486 display shield on a Nordic nRF52840-DK board. The other modes are untested.

Any comments or feedback are appreciated.

stgloorious committed:
drivers: display: ili9xxx: read mipi-mode, only use SPI mode as fallback

Instead of always using the SPI MIPI DBI mode (type C), look up the
mipi-mode from the device tree and only set the mode to
MIPI_DBI_MODE_SPI_4WIRE as a fallback in case the property is not given.

Signed-off-by: Stefan Gloor <code@stefan-gloor.ch>

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_gpio.c:
```diff
@@ -0,0 +1,373 @@
+			/* Calculate a look-up table to support arbitrary bit orders. */
+			for (int i = 0; i < (1 << config->data_bus_width); i++) {
+				data->data_lut[i] = 0;
+				for (int j = 0; j < config->data_bus_width; j++) {
```
I *think* we could generate this LUT statically like so:

```
#define LUT_GEN(i, _) ((i & (1 << 0)) ? DT_INST_GPIO_PIN_BY_IDX(n, data_gpios, 0) : 0) \
                        ((i & (1 << 1)) ? DT_INST_GPIO_PIN_BY_IDX(n, data_gpios, 1) : 0) \
                        ((i & (1 << 2)) ? DT_INST_GPIO_PIN_BY_IDX(n, data_gpios, 2) : 0)
                        /* ... continue up to 8 wide data bus ... */

uint32_t lut[] = {
        LISTIFY(256, LUT_GEN, (,))
};
```

Not proud of these macrobatics, but I think it might allow us to save the 1Kib of RAM here.

stgloorious replied:
Done.

danieldegrasse reviewed (commented):
Very cool, thanks for this! Please take a look at the compliance failure when you get a chance (you can amend commits and force push over this branch to respond to review requests/fix compliance issues)

![image](https://github.com/user-attachments/assets/8d2ccd6d-6c7b-4f6d-b148-03fa57607074)

stgloorious committed:
doc: update MIPI DBI API compatibility

The MIPI DBI API supports MIPI DBI controllers type A, B, and C
(except with 16 write clocks). Update the documentation accordingly.

Signed-off-by: Stefan Gloor <code@stefan-gloor.ch>

stgloorious commented on drivers/display/display_ili9xxx.c:
```diff
@@ -520,7 +520,8 @@ static const struct ili9xxx_quirks ili9488_quirks = {
 		.mipi_dev = DEVICE_DT_GET(DT_PARENT(INST_DT_ILI9XXX(n, t))),   \
 		.dbi_config = {                                                \
-			.mode = MIPI_DBI_MODE_SPI_4WIRE,                       \
+			.mode = DT_PROP_OR(INST_DT_ILI9XXX(n, t),              \
```
I subscribed to that thread and will remove my commit once it is merged.

stgloorious commented on drivers/mipi_dbi/Kconfig.gpio:
```diff
@@ -0,0 +1,12 @@
+# Copyright 2024 Stefan Gloor
+# SPDX-License-Identifier: Apache-2.0
+
+config MIPI_DBI_GPIO
```
Done.

stgloorious commented on drivers/mipi_dbi/mipi_dbi_gpio.c:
```diff
@@ -0,0 +1,373 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mipi_dbi_gpio, CONFIG_MIPI_DBI_LOG_LEVEL);
+
+/* Although the MIPI DBI spec only allows 8, 9 and 16 bits, we support up to 32 bit wide buses. */
```
I thought I had seen those before but I think I start to confuse different interfaces... I haven't found any MIPI DBI examples in a quick search. Better to stick to the standard then.

stgloorious commented on drivers/mipi_dbi/mipi_dbi_gpio.c:
```diff
@@ -0,0 +1,373 @@
+	const struct device *port;
+
+	/* Data GPIO remap look-up table. Valid if single_port is set */
+	uint32_t data_lut[256];
```
Very cool, I didn't know about `LISTIFY`.

stgloorious commented on doc/hardware/peripherals/mipi_dbi.rst:
```diff
@@ -20,7 +20,8 @@ MIPI DBI defines 3 interface types:
   #. 8 write clocks per byte. Command/data selected via GPIO pin
 
-Currently, the API only supports Type C controllers, options 1 and 3.
+Currently, the API does not support Type C controllers with 16 write clocks
```
Done.

stgloorious replied:
The same information is also in the API header: `include/zephyr/drivers/mipi_dbi.h`. I updated it as well.

stgloorious commented on drivers/mipi_dbi/mipi_dbi_gpio.c:
```diff
@@ -0,0 +1,373 @@
+	return mipi_dbi_gpio_write_helper(dev, dbi_config, false, 0x0, framebuf, desc->buf_size);
+}
+
+static int mipi_dbi_gpio_command_read(const struct device *dev,
```
You're right, removed it.

stgloorious commented on drivers/mipi_dbi/mipi_dbi_gpio.c:
```diff
@@ -0,0 +1,373 @@
+	return gpio_pin_set_dt(&config->reset, 0);
+}
+
+static int mipi_dbi_gpio_release(const struct device *dev, const struct mipi_dbi_config *dbi_config)
```
You're right, removed it.

danieldegrasse commented on include/zephyr/drivers/mipi_dbi.h:
```diff
@@ -15,7 +15,7 @@
  *     2. Same as above, but 16 write clocks per byte
  *     3. 8 write clocks per byte. Command/data selected via GPIO pin
- * The current driver interface only supports type C modes 1 and 3
+ * The current driver interface does not support type C with 16 write clocks (option 2).
```
This change should be moved to the prior commit

stgloorious replied:
Done.

danieldegrasse commented on drivers/mipi_dbi/mipi_dbi_bitbang.c:
```diff
@@ -0,0 +1,351 @@
+	return mipi_dbi_bitbang_write_helper(dev, dbi_config, false, 0x0, framebuf, desc->buf_size);
+}
+
+static inline bool mipi_dbi_has_pin(const struct gpio_dt_spec *spec)
```
Is this used? Looks like we can drop it

stgloorious replied:
No, dropped.

stgloorious commented:
Hi @danieldegrasse, thank you very much for your review. I should have addressed all the issues mentioned by you or the CI. I'll update again if there are still CI problems once it runs through.

Let me know what you think about the changes and if you think anything else could be improved.

stgloorious committed:
drivers: mipi_dbi: add support for parallel 8080/6800 modes using GPIO

Introduce GPIO-based driver for MIPI DBI class that allows MIPI DBI
type A and B displays to be used on general platforms.

Since each data pin GPIO can be selected individually, the bus pins are
set in a loop, which has a significant negative impact on performance.
When using 8-bit mode and all the data GPIO pins are on the same port,
a look-up table is generated to set the whole port at once as a
performance optimization. This creates a ROM overhead of about 1 kiB.

Tested 8-bit 8080 mode with ILI9486 display on nRF52840-DK board.

Signed-off-by: Stefan Gloor <code@stefan-gloor.ch>

stgloorious commented:
![screenshot-Mo 16 Sep 2024 22:29:40 CEST](https://github.com/user-attachments/assets/2f9f9f0a-bd09-4ce8-b226-97197ad0e3a6)
 
The CI doesn't like the documentation change. What am I doing wrong?

stgloorious commented:
It failed [again](https://github.com/zephyrproject-rtos/zephyr/actions/runs/10891696880/job/30223149597) with the new forced push in [check-valid-pr](https://github.com/carpentries/actions/blob/v0.14.0/check-valid-pr/index.js#L61). For some reason it checks  the old commit 30aa0d455262d643d0ab6e01f7db729ffc347d45 against the current HEAD 61130c72ed6756773981b55458e7d0fe3712c138, and that seems to fail because probably  `headroom <= 1`.  Maybe just rerunning helps?

danieldegrasse commented:
> It failed [again](https://github.com/zephyrproject-rtos/zephyr/actions/runs/10891696880/job/30223149597) with the new forced push in [check-valid-pr](https://github.com/carpentries/actions/blob/v0.14.0/check-valid-pr/index.js#L61). For some reason it checks the old commit [30aa0d4](https://github.com/zephyrproject-rtos/zephyr/commit/30aa0d455262d643d0ab6e01f7db729ffc347d45) against the current HEAD [61130c7](https://github.com/zephyrproject-rtos/zephyr/commit/61130c72ed6756773981b55458e7d0fe3712c138), and that seems to fail because probably `headroom <= 1`. Maybe just rerunning helps?

That is odd, I haven't seen this check fail before. I'll investigate if it fails again there
